### ISSUE TITLE
PHP Warning - The DateTimeInterface object has not been correctly initialized by its constructor

### ISSUE THREAD

**Issue by paulopereiraptc** [igbinary/igbinary#103]
Hello I'm getting an error using the DateTime structures with igbinary (tested with v1.2.1 and v2.0.0) session_handle on php 7.0.13. The following script shows a different output using the php 7 and php 5.6:

```
<?php

// Start session
session_start();
$_SESSION['date'] = new \DateTime();
echo "\n";
$mydata = session_encode();
$sessionData = session_decode($mydata);
var_dump($_SESSION['date']->diff(new \DateTime())->days);
```

Output from php 7.0.13:
```
PHP Warning:  DateTime::diff(): The DateTimeInterface object has not been correctly initialized by its constructor in /root/serializer_test.php on line 21
PHP Notice:  Trying to get property of non-object in /root/serializer_test.php on line 21
NULL
```

Output from 5.6.18:
```
int(0)
```

Both versions have the variable "session.serialize_handler=igbinary". When I changed this variable to "php" the warning disappears and  the php 7 returns the result with success. The funny thing is this problem doesn't appear when we only use the "igbinary_serialize" and "igbinary_unserialize" functions:

```
<?php

$data = igbinary_serialize(['date'=>new \DateTime()]);

$dataArray = igbinary_unserialize($data);

var_dump($dataArray['date']->diff(new \DateTime())->days);
```

Output from php 7.0.13:
```
int(0)
```

Output from 5.6.18:
```
int(0)
```

 Any idea how to fix this problem?

Thanks in advance :)

**Comment by TysonAndre** [comment]
Which serialize save handler?

Does this happen for CLI invocations or just on a web server?

What's the value of `bin2hex(var_export($_SESSION))`

**Comment by paulopereiraptc** [comment]
Hello,

Right now the error happens in CLI and FPM. I run the following code in the CLI:

```
<?php
// Start session
session_start();
$_SESSION['date'] = new \DateTime();
var_dump(bin2hex(igbinary_serialize($_SESSION)));
```

Output from php 7.0.13:
```
string(172) "00000002140111046461746517084461746554696d6514030e00111a323031362d31312d31362031373a33383a31322e303030303030110d74696d657a6f6e655f747970650603110874696d657a6f6e651103555443"
```

Output from 5.6.18:
```
string(192) "00000002140111046461746517084461746554696d6514030e00111a323031362d31312d31362031373a33393a32372e303030303030110d74696d657a6f6e655f747970650603110874696d657a6f6e65110d4575726f70652f4c6973626f6e"
```

Thanks :)

**Comment by paulopereiraptc** [comment]
Example using the same timestamp:

```
<?php
// Start session
session_start();
$_SESSION['date'] = new \DateTime('2016-11-16');
var_dump(bin2hex(igbinary_serialize($_SESSION)));
```

Output from php 7.0.13:
```
string(192) "00000002140111046461746517084461746554696d6514030e00111a323031362d31312d31362030303a30303a30302e303030303030110d74696d657a6f6e655f747970650603110874696d657a6f6e65110d4575726f70652f4c6973626f6e"
```

Output from 5.6.18:
```
string(192) "00000002140111046461746517084461746554696d6514030e00111a323031362d31312d31362030303a30303a30302e303030303030110d74696d657a6f6e655f747970650603110874696d657a6f6e65110d4575726f70652f4c6973626f6e"
```

**Comment by paulopereiraptc** [comment]
Using the function "session_encode" instead of "igbinary_serialize" the result is the same:

```
<?php

// Start session
session_start();

$_SESSION['date'] = new \DateTime('2016-11-16');

var_dump(bin2hex(session_encode()));
```

Output from php 7.0.13:

```
string(192) "00000002140111046461746517084461746554696d6514030e00111a323031362d31312d31362030303a30303a30302e303030303030110d74696d657a6f6e655f747970650603110874696d657a6f6e65110d4575726f70652f4c6973626f6e"
```

Output from 5.6.18:
```
string(192) "00000002140111046461746517084461746554696d6514030e00111a323031362d31312d31362030303a30303a30302e303030303030110d74696d657a6f6e655f747970650603110874696d657a6f6e65110d4575726f70652f4c6973626f6e"
```

**Comment by TysonAndre** [comment]
I'm able to reproduce this in 7.0.13.

The php developers changed the way serialize decoders are supposed to add fields to the global $_SESSION and how reference counting was done, in order to fix other bugs.

This may take a while to test and to identify affected versions.

**Comment by paulopereiraptc** [comment]
Hello,

It's possible the problem happen with all sub-versions of php 7 (I test with php 7.0.0 and I got the same problem ...). In the php 5.6 I'm not able to reproduce the problem ...

**Comment by TysonAndre** [comment]
This is fixed on master, and will be released in igbinary 2.0.1 happening because igbinary wasn't calling __wakeup on objects in the session decoder API (Only two top level APIs needed to do that).

Only php 7 session decoding was affected.
